source/Ship.cpp

```diff
@@ -3788,20 +3804,101 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
 // cues and try to stay with it when it lands or goes into hyperspace.
 void Ship::AddEscort(Ship &ship)
 {
-	escorts.push_back(ship.shared_from_this());
+	auto sp = ship.shared_from_this();
+	TuneForEscort(sp);
+	escorts.list.push_back(std::move(sp));
 }
 
 
 
 void Ship::RemoveEscort(const Ship &ship)
 {
-	auto it = escorts.begin();
-	for( ; it != escorts.end(); ++it)
-		if(it->lock().get() == &ship)
+	// Reset the cached value if we are removing the slowest escort, we will
+	// re-scan the list of escorts to set the escorts velocity based on
+	// remaining active escorts.
+	shared_ptr<Ship> slowest = escorts.slowest.lock();
+	bool findSlowest = (!slowest || (&ship == slowest.get()));
+	if(findSlowest){
+		escorts.cruiseVelocity = -1.;
+		escorts.slowest.reset();
+	}
+
+	auto it = escorts.list.begin();
+	while(it != escorts.list.end())
+	{
+		auto escort = it->lock();
+		if(escort.get() == &ship)
 		{
-			escorts.erase(it);
-			return;
+			it = escorts.list.erase(it);
+			// If we are not removing the slowest escort, then we don't
+			// need to finish the loop to tune all other escorts.
+			if(!findSlowest)
+				return;
+		}
+		else
+		{
+			if(escort && findSlowest)
+				TuneForEscort(escort);
+			++it;
+		}
+	}
+}
+
+
+
+// (Re)Register escorts, for example because some escort got destroyed.
+void Ship::TuneForEscorts()
+{
+	// Check if we have a valid slowest escort. If we have one, then we don't
+	// need to rescan for the next slowest escort.
+	shared_ptr<Ship> slowest = escorts.slowest.lock();
+	if(slowest && !slowest->IsDestroyed() && government == slowest->GetGovernment())
+		return;
+
+	// Reset this value, we will re-scan the list of escorts to set the
+	// escorts velocity based on remaining active escorts.
```

Comment:
This is a "what" comment rather than a "why," and doesn't offer anything the preceding comment or the invoked method comment(s) dont already offer.
```suggestion
```